ports/nrf/common-hal/_bleio/bonding.c

```diff
@@ -0,0 +1,312 @@
+    bonding_block_t *existing_block =
+        find_existing_block(connection->is_central, BLOCK_SYS_ATTR, connection->ediv);
+    if (existing_block) {
+        if (length == existing_block->data_length &&
+            memcmp(sys_attr, existing_block->data, length) == 0) {
+            // Identical block found. No need to store again.
+            return;
+        }
+        // Data doesn't match. Invalidate block and store a new one.
+        invalidate_block(existing_block);
+    }
+
+    bonding_block_t block_header = {
+        .is_central = connection->is_central,
+        .type = BLOCK_SYS_ATTR,
+        .ediv = connection->ediv,
+        .conn_handle = connection->conn_handle,
+        .data_length = length,
+    };
+    bonding_block_t *new_block = find_unused_block(length);
+    write_block_header(new_block, &block_header);
+    write_block_data(new_block, sys_attr, length);
+    return;
+}
+
+STATIC void write_keys_block(bleio_connection_internal_t *connection) {
+    uint16_t const ediv = connection->is_central
+        ? connection->bonding_keys.peer_enc.master_id.ediv
+        : connection->bonding_keys.own_enc.master_id.ediv;
+
+    // Is there an existing keys block that matches?
+    bonding_block_t *existing_block = find_existing_block(connection->is_central, BLOCK_KEYS, ediv);
+    if (existing_block) {
+        if (existing_block->data_length == sizeof(bonding_keys_t) &&
+            memcmp(existing_block->data, &connection->bonding_keys, sizeof(bonding_keys_t)) == 0) {
+            // Identical block found. No need to store again.
+            return;
+        }
+        // Data doesn't match. Invalidate block and store a new one.
+        invalidate_block(existing_block);
+    }
+
+    bonding_block_t block_header = {
+        .is_central = connection->is_central,
+        .type = BLOCK_KEYS,
+        .ediv = ediv,
+        .conn_handle = connection->conn_handle,
+        .data_length = sizeof(bonding_keys_t),
+    };
+    bonding_block_t *new_block = find_unused_block(sizeof(bonding_keys_t));
+    write_block_header(new_block, &block_header);
+    write_block_data(new_block, (uint8_t *) &connection->bonding_keys, sizeof(bonding_keys_t));
+}
+
+void bonding_clear_keys(bonding_keys_t *bonding_keys) {
+    memset((uint8_t*) bonding_keys, 0, sizeof(bonding_keys_t));
+}
+
+void bonding_reset(void) {
+    if (BONDING_FLAG != *((uint32_t *) BONDING_START_FLAG_ADDR) ||
+        BONDING_FLAG != *((uint32_t *) BONDING_END_FLAG_ADDR)) {
+        bonding_erase_storage();
+    }
+}
+
+// Write bonding blocks to flash. Requests have been queued during evt handlers.
+void bonding_background(void) {
+    // A paired connection will request that its keys and CCCD values be stored.
+    // The CCCD store  whenever a CCCD value is written.
+    for (size_t i = 0; i < BLEIO_TOTAL_CONNECTION_COUNT; i++) {
+        bleio_connection_internal_t *connection = &bleio_connections[i];
+
```

Comment:
Do you want to save bonding info when the connection has already been lost? (The handle is invalid.) There may need to be some locking here too since connection state can change in interrupts.

Reply:
I think this is safe, because the `sd_ble_sys_attr_get()` calls in `write_sys_attr_block()` will fail if the connection handle is invalid, and then `write_sys_attr_block()` will just give up in that case. 

Interestingly the documentation talks about (ideally) fetching the values after the connection is closed to avoid churn. But someone may press reset, etc. So that's why I put in the one-second delay, which seems to cover the typical case.


> uint32_t sd_ble_gatts_sys_attr_get | ( | uint16_t | conn_handle,
> -- | -- | -- | --
>   |   | uint8_t * | p_sys_attr_data,
>   |   | uint16_t * | p_len,
>   |   | uint32_t | flags
>   | ) |   |  
> 
> Retrieve persistent system attribute information from the stack.
> 
> This call is used to retrieve information about values to be stored persistently by the application during the lifetime of a connection or after it has been terminated. When a new connection is established with the same bonded device, the system attribute information retrieved with this function should be restored using using sd_ble_gatts_sys_attr_set. If retrieved after disconnection, the data should be read before a new connection established. The connection handle for the previous, now disconnected, connection will remain valid until a new one is created to allow this API call to refer to it. Connection handles belonging to active connections can be used as well, but care should be taken since the system attributes may be written to at any time by the peer during a connection's lifetime.